### ISSUE TITLE
Get or create an AppCredential for an AAD application

### ISSUE THREAD

**Issue by wangzq** [Azure/azure-powershell#3012]
### Cmdlet(s)
Get-AzureRmAdAppCredential, New-AzureRmAdAppCredential

### PowerShell Version
5.1.14393.187

### Module Version
2.0.1

### OS Version
10.0.14393.187

### Description
I want to first retrieve existing AppCredentials for an application then determine if a specific certificate is already uploaded, so that I can only upload it when it is not there. However, since I cannot specify the KeyId when uploading, and the returned data doesn't contain the CertValue, it is not possible to achieve this. I tried to use StartDate/EndDate to compare with cert.NotBefore and NotAfter as a workaround, and found they are returned as strings instead of datetime, which also make this not easy.

**Comment by markcowl** [comment]
@shuagarw can you take a look?

**Comment by aiden56** [comment]
*Following* The current workaround is to not use this - but instead use the Msol cmdlets to achieve an idempotent result.

**Comment by DirtyPiece** [comment]
I would also love to see something like this.

**Comment by wangzq** [comment]
Is it possible to return the thumbprint as part of the returned credential object from Get-AzureRmAdAppCredential?

**Comment by wangzq** [comment]
I noticed the KeyCredential object does have a property "customKeyIdentifier": https://msdn.microsoft.com/en-us/library/azure/ad/graph/api/entity-and-complex-type-reference#keycredential-type

And searching shows this might be designed for this purpose, such as storing a base64 encoded thumbprint in this property: https://blogs.msdn.microsoft.com/mrochon/2016/05/09/azure-ad-client-certificate-rollover/

Therefore, I think an easier fix for this issue, is to ensure when we use New-AzureRmAdAppCredential, we use the thumbprint to fill in this customKeyIdentifier (or better, let caller decide to pass in a value for it); and in the returned object from Get-AzureRmAdAppCredential, we simply include this customKeyIdentifier property, which I can see it is already in the Graph API returned JSON, it is just currently Azure-PowerShell discards it.

I can work on a PR if you agree with this. 

Thanks

**Comment by shuagarw** [comment]
@danieldobalian for FYI.
It seems fine to me. Interesting that graph documentation tells "customKeyIdentifier" is of type "Edm.Binary".

**Comment by degant** [comment]
Yes this seems like a very useful feature. We just encountered the same and tried using Start Date & End Date for comparison as a workaround. 

Would be great to have the ability to check thumbprints of all uploaded certificates on an AAD resource, preferably as a member in the response for Get-AzureRmADAppCredential.

**Comment by andreesteve** [comment]
@markcowl - do you know if this is planned for a future release? This feature would allow us to move completely out of AzureAD module into Az !

**Comment by henning-krause** [comment]
I would also like to have this feature.

**Comment by MaddyMicrosoft** [comment]
The AzureRM module is retired. The asked-for features have been implemented on New-AzADAppCredential and Get-AzADAppCredential in Az.Resources (PR #14935, shipped in v4.1.0). Closing as complete.
